src/base64.cr

```diff
@@ -110,36 +113,49 @@ module Base64
   end
 
   private def from_base64(data)
-    buf = 0
-    mod = 0
+    len = data.length
     dt = DECODE_TABLE.buffer
-    data.each do |byte|
-      dec = dt[byte]
-      if dec < 0
-        next if dec == -2
-        break if dec == -3
-        raise Error.new("Invalid character '#{byte.chr}'")
-      end
-      buf = (buf | dec) << 6
-      mod += 1
-      if mod == 4
-        mod = 0
-        yield (buf >> 22).to_u8
-        yield (buf >> 14).to_u8
-        yield (buf >> 6).to_u8
+    cstr = data.pointer(len)
+    while (sym = cstr[len - 1]) && (sym == NL || sym == NR || sym == PAD) && (len > 0)
+      len -= 1
+    end
+    endcstr = cstr + len - 4
+
+    while cstr <= endcstr
+      a, b, c, d = next_decoded_value, next_decoded_value, next_decoded_value, next_decoded_value
+
+      yield (a << 2 | b >> 4).to_u8
+      yield (b << 4 | c >> 2).to_u8
+      yield (c << 6 | d).to_u8
+
+      while (cstr.value == NL || cstr.value == NR) && cstr <= endcstr
+        cstr += 1
       end
     end
 
+    mod = (endcstr - cstr) % 4
     if mod == 2
-      yield (buf >> 10).to_u8
+      a, b = next_decoded_value, next_decoded_value
+      yield (a << 2 | b >> 4).to_u8
     elsif mod == 3
-      yield (buf >> 16).to_u8
-      yield (buf >> 8).to_u8
+      a, b, c = next_decoded_value, next_decoded_value, next_decoded_value
+      yield (a << 2 | b >> 4).to_u8
+      yield (b << 4 | c >> 2).to_u8
     elsif mod != 0
       raise Error.new("Wrong length")
     end
   end
 
+  private macro next_decoded_value
```

Comment:
For that we would need some sort of escape analysis. Or maybe for now we could track the variable that was assigned a closure and see if it's only invoked and never reassigned or passed to a method. But these are optimizations that I don't think we'll do soon.

Another alternative would be to allow defining nested defs, but that's something new too.

I think macros are perfectly fine for what you did. For example one thing that you can't do with normal methods is to return early from the parent caller. For example:

```crystal
def foo
  check_some_condition
end

def check_some_condition
  if some_condition
    # But I want to return from foo here!
    return
  end
end
```

With macros this is possible:

```crystal
def foo
  check_some_condition
end

macro check_some_condition
  if some_condition
    # But I want to return from foo here!
    return
  end
end
```

Also, in these cases the macro is just local to that class, so it's just a helper. Accessing outside variables like that isn't a big deal. Also, if you accidentally use incorrect names you will get a compile error, so it's safe.